controller/appcontroller.go

```diff
@@ -1389,15 +1388,43 @@ func (ctrl *ApplicationController) processAppRefreshQueueItem() (processNext boo
 		localManifests = opState.Operation.Sync.Manifests
 	}
 
-	revision := app.Spec.Source.TargetRevision
-	if comparisonLevel == CompareWithRecent {
-		revision = app.Status.Sync.Revision
-	}
+	revisions := make([]string, 0)
+	sources := make([]appv1.ApplicationSource, 0)
+	sourceMap := make(map[string]int)
+
+	hasMultipleSources := app.Spec.Sources != nil && len(app.Spec.Sources) > 0
 
+	// If we have multiple sources, we use all the sources under `sources` field and ignore source under `source` field.
+	// else we use the source under the source field.
+	if hasMultipleSources {
+		for _, source := range app.Spec.Sources {
+			// filter out repeated sources having same repoUrl
+			if i, ok := sourceMap[source.RepoURL]; ok {
```

Comment:
I think it is valid use case to use the same repo URL and two different paths. Because of this check the controller logic is inconsistent. Seems like sync is working correctly but diffing wants to delete second source (UI is showing and second source should be deleted):

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
  namespace: argocd
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: default
  sources:
  - path: guestbook
    repoURL: https://github.com/argoproj/argocd-example-apps
    targetRevision: HEAD
  - path: kustomize-guestbook
    repoURL: https://github.com/argoproj/argocd-example-apps
    targetRevision: HEAD
```

Reply:
As per our discussion in slack, removed dedup checks and will let ArgoCD handle the duplicate resources.